### PR TITLE
JSON feedback

### PR DESCRIPTION
**Pull request by Jordie0608**
Replaces the string and int based feedback system with one that records data entirely in JSONs (LONGTEXT in Maria). For how this works see: https://github.com/Jordie0608/tgstation/blob/6ef229ddc4aa227b2f911def66b716bfa572fca2/code/controllers/subsystem/blackbox.dm#L129-L182
The blackbox subsystem no longer stores every radio message sent just to count how many there were and is updated to identify the current radio channels. The subsystem now stores the highest tech levels attained since everywhere else this is stored can be lost during the round, deleting the data.

Some feedback keys have been changed or removed, a spreadsheet detailing all of these changes and notes pertaining to them can be found at: https://file.house/q7nI.ods

DM side functionality tested to work. A couple of stat related fixes are scattered in, most being incorrect arguments being used.

Conversion of old data is done by a python script. As my first time in python it's not very pretty but I've tested that it handles all our feedback data as of 2017-10-23 01:34:06. Details on how it operates are in the script itself

@nfreader 
@MrStonedOne

### CODE REVIEW

**Comment by nfreader** [comment]
It... BEGINS!

**Comment by Jordie0608** [comment]
Making the conversion query is going to be a bit delayed while I learn python for it.

**Comment by optimumtact** [comment]
wouldn't that make it a script

File: code/controllers/subsystem/blackbox.dm [hunk 1]
```diff
@@ -71,112 +48,195 @@ SUBSYSTEM_DEF(blackbox)
...
+/datum/controller/subsystem/blackbox/proc/log_research(tech, level)
+	if(!(tech in research_levels) || research_levels[tech] < level)
+		research_levels[tech] = level
 
-/datum/controller/subsystem/blackbox/proc/LogBroadcast(blackbox_msg, freq)
+/datum/controller/subsystem/blackbox/proc/LogBroadcast(freq)
 	if(sealed)
 		return
 	switch(freq)
 		if(1459)
```
Comment (lzimann) [inline code/controllers/subsystem/blackbox.dm]: no const for common?

Comment (Jordie0608) [reply code/controllers/subsystem/blackbox.dm]: Nope https://github.com/tgstation/tgstation/blob/master/code/game/communications.dm#L158

**Comment by Jordie0608** [comment]
Python script finished finally, now for testing.

File: code/controllers/subsystem/blackbox.dm [hunk 2]
```diff
@@ -71,112 +48,197 @@ SUBSYSTEM_DEF(blackbox)
...
-	var/datum/feedback_variable/FV = find_feedback_datum(variable)
-	FV.set_value(value)
-
-/datum/controller/subsystem/blackbox/proc/inc(variable, value)
-	if(sealed)
-		return
-	var/datum/feedback_variable/FV = find_feedback_datum(variable)
-	FV.inc(value)
+	var/datum/feedback_variable/FV = find_feedback_datum(key, type)
+	switch(type)
```
Comment (lzimann) [inline code/controllers/subsystem/blackbox.dm]: The type could use a define.

**Comment by Jordie0608** [comment]
I'm going to do some paranoia testing tomorrow but this is all but done.

**Comment by nfreader** [comment]
#herewego

**Comment by optimumtact** [comment]
@MrStonedOne can you coordinate on the merge of this

**Comment by Jordie0608** [comment]
More testing done, still potentially edge cases somewhere due to typos but otherwise we're good.
capture

@optimumtact It will require MSO

**Comment by optimumtact** [comment]
alright, lets try to arrange all the people

**Comment by optimumtact** [comment]
@Jordie0608 why did the number of rows drop so much? Are you combining some of them into array structures?

**Comment by Jordie0608** [comment]
Yes and some are also removed from current tracking or obsolete and no longer tracked at all.

**Comment by JJRcop** [comment]
Did the database get updated when this was merged?

**Comment by Jordie0608** [comment]
Yes

**Comment by nfreader** [comment]
Statnarok!!!!
